### ISSUE TITLE
Generic arguments/returns of specialized subclass type report incompatibility with superclass's Self type

### ISSUE THREAD

**Issue by BaconPancakes** [python/mypy#16165]
**Bug Report**

I want to define methods for my generic type that implement a boolean interface, only when the TypeVar is bool. It looks type safe to me but mypy reports incompatible types.

**To Reproduce**

```python
from typing import Protocol, Self, TypeVar, Generic

T = TypeVar("T")

class Base(Protocol):
    def __and__(self: Self, other:Self) -> Self:
        pass
    
class Derived(Generic[T], Base):
    def __and__(self: Derived[bool], other: Derived[bool]) -> Derived[bool]:
        return Derived[bool]()
```
https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=4c33bbfd24b5d46ff80801ca26dc710d

**Expected Behavior**

I think this should be allowed. If you instantiate with anything other than a `Derived[bool]` then the `__and__` operator should report an error. For example I'd expect
```python
d1 = Derived[int]()
d2 = d1
result = d2 & d1  # type error should emit

d3 = Derived[bool]()
d4 = d3

result = d3 & d4  # no type error should occur
```
`pyright` seems to handle this all correctly.

**Actual Behavior**

```
main.py:10: error: Return type "Derived[bool]" of "__and__" incompatible with return type "Derived[T]" in supertype "Base"  [override]
main.py:10: error: Argument 1 of "__and__" is incompatible with supertype "Base"; supertype defines the argument type as "Derived[T]"  [override]
main.py:10: note: This violates the Liskov substitution principle
main.py:10: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
Found 2 errors in 1 file (checked 1 source file)

**Comment by JelleZijlstra** [comment]
I don't think this should be allowed. The Protocol requires that `__and__` is supported for all instances, and by explicitly inheriting from `Base` you declare that you fulfill the Protocol.

The fact that pyright doesn't report this error feels like a bug, cc @erictraut.

**Comment by erictraut** [comment]
Yes, I agree this is a bug in pyright. Mypy is correct.

**Comment by BaconPancakes** [comment]
Thanks for the response, folks.

> The Protocol requires that __and__ is supported for all instances

Can it not be true that only the `Derived[bool]` satisfies the `Protocol` but generally not necessarily all `Derived[T]`?

It seems to work out when you don't explicitly inherit from the protocol. Is the following expected?

```python
from typing import Protocol, Self, TypeVar, Generic

T = TypeVar("T")

class Base(Protocol):
    def __and__(self: Self, other:Self) -> Self:
        pass
    
class Derived(Generic[T], Base):
    def __and__(self: Derived[bool], other: Derived[bool]) -> Derived[bool]:
        return Derived[bool]()
        

class ImplicitDerived(Generic[T]):
    def __and__(self: ImplicitDerived[bool], other: ImplicitDerived[bool]) -> ImplicitDerived[bool]:
        return ImplicitDerived[bool]()
        
class NotDerived():
    pass
        
        
def fun(b: Base) -> bool:
    return True
    

r = fun(NotDerived())  # error: Argument 1 to "fun" has incompatible type "NotDerived"; expected "Base"  [arg-type]

# A class object is considered an implementation of a protocol if accessing all members on it results in types compatible with the protocol members.
r = fun(ImplicitDerived[bool]())  # passes: ImplicitDerived[bool] satisfied Base
```
https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=7d51574491006a6a086c74470feb03d7

**Comment by erictraut** [comment]
Yes, this is expected. Note that `fun(ImplicitDerived[str]())` results in an error.

`ImplicitDerived[bool]` is compatible with `Base`, but `ImplicitDerived[T]` is not because the protocol stipulates that the `__and__` method must work with all values of type `Self`.

**Comment by BaconPancakes** [comment]
That’s pretty much what I wanted out of the API, which is good. I presume trying to make this relationship explicit (for the benefit of class definition static checks) like I want in the original `Derived[T]` is out of scope of the typing specification?
